### ISSUE TITLE
`aria-owns` support

### ISSUE THREAD

**Issue by bbudd** [nvaccess/nvda#8439]
I apologize for not following the template, but this is more of a feature request than a bug:

I notice that your `aria-owns` does not work except in chrome—and I think that that is incidental and more to do with chrome just working with a lot of readers more than a specific attempt on your part to work with the spec.

https://www.w3.org/TR/wai-aria-1.1/#aria-owns

**Comment by derekriemer** [comment]
Not enough information is provided here.

**Comment by leonardder** [comment]
@bbudd: I tend to disagree on your saying that the issue template doesn't fit your case. If you want your request to be valid, please fill in the issue template as requested (i.e. provide the steps to reproduce your finding that aria-owns doesn't do what you expect it to do).

**Comment by josephsl** [comment]
Hi,

I'll open another issue that discusses this further, but to elaborate what @DerekRiemer (who works at Google these days) said:

When submitting a feature request, please give us a clearer picture. A "clearer picture" is the one that is easy for you and us (developers) to understand, namely including a rationale as to why the feature should be included in NVDA Core, and content that demonstrates that you thought about it. Just stating that "... NvDA should support such and such because the spec says it ..." is, to me (and I'm sure for others), not enough justification to take your request more seriously.

Also, to elaborate on @LeonardDer's remark: sometimes, when you think carefully, a feature request can be done in a way that lets you use the issue template. For example, "expected versus actual" sections are useful to demonstrate what NVDA does not do now versus what it can do (I'm intentionally avoiding the word "should").

Thanks.

**Comment by bbudd** [comment]
I'm sorry, but I disagree that "because the spec says it should" isn't enough rationale to support something. That's why specs exists, to give engineers an itemized list of functionality to support and guidelines for how specifically to support it. Otherwise you're just building what you want, and nothing else needs matter. *Especially* in the world of accessibility, where proper coverage is spotty at best.

Saying "'you don't yet support this part of the spec' isn't enough information," is being needlessly pedantic. Yes, I could have filled out the template, but it'd be full of generalizations, like "expected behavior: `aria-owns` should work; actual behavior: `aria-owns` doesn't work," which isn't any real way to understand the issue any more (which is the purpose templates like those serve). Provide steps? Come on. You have missing functionality, and you need me to provide steps to find it? I'm more than willing to provide you steps to a specific bug inside your codebase should the need arise. But pointing out to a car builder that they're missing headlights doesn't require steps (plural), it requires one step - look at the car. 

Look, I've been a software engineer for 20 years, and I understand the benefit of having issue templates, I truly do. But pointing out missing functionality shouldn't ever require filling out a template meant for bugs. Maybe there's a better solution for requesting and tracking features than your issue board? I'm actually only requesting this so I can track it, not because I need it built today.

**Comment by leonardder** [comment]
[**@bbudd**](https://github.com/bbudd) commented on [26 jun. 2018 16:09 CEST](https://github.com/nvaccess/nvda/issues/8439#issuecomment-400322504 "2018-06-26T14:09:06Z - Replied by Github Reply Comments"):
> Yes, I could have filled out the template, but it'd be full of generalizations, like "expected behavior: `aria-owns` should work; actual behavior: `aria-owns` doesn't work," which isn't any real way to understand the issue any more (which is the purpose templates like those serve). 

I agree that "aria-owns should work", "aria-owns doesn't work" doesn't make sense, but that's not at all what we're asking for here. The question I'm asking and the question the template is asking is: what do you expect, and what is happening instead? What is NVDA supposed to do when aria-owns works (i.e. how should it behave), and what does NVDA do what makes you say that it doesn't work?

**Comment by josephsl** [comment]
Hi,

Regarding specs and such: I do agree that specs can help us (engineers) define a specific product and goals to achieve something. However, I think it might not be a good idea to just judge a product because it does not follow specific specs:

1. Specs cover common cases: I'm sure we can agree that specs cover common ground when it comes to tackling an issue. However, from what I can tell (based on issues on GitHub and reports from users), many run into situations that are not covered by baseline specs, which means AT's are free to do whatever it wishes for corner cases.
2. Human factors: specs serve as directives to engineers as to how to do things via technical means. But that's only one part of the story: users matter, and in extension, human factors become crucial when it comes to providing meaningful experience. While some specs do allow useful experiences to be opened up, some were not taken up by AT's due to impact on users, especially productivity, and some are controversial in developer circles.
3. Different interpretations: because we are humans, we have different interpretations about specs. A spec that appeals to certain audiences may not be so for others. Some may have differing interpretations as to how to implement small things that is actually going to have biggest impact. Also, interpretation differences are not just common among engineers: it is seen among websites, hence even when requested to add ARIA attributes , not all websites follow guidelines and may have differing implementations.

UI Automation (UIA) is a good analogy for this case. Just because UIA says AT's should announce certain things does not mean it is taken up by screen reader vendors, nor should be a major criteria for judging products. For example, NVDA does not blindly announce live region changes sent by UIA all the time - there are cases where it should not because it is detrimental to user experience or bad UIA implementations.

As for aria-owns question: @JCSTeh, do you have any thoughts on all this? I'd like to ask you because you've worked on aria specs and now works for Mozilla so you'd know how web browsers and AT's should cooperate.

Thanks.

**Comment by derekriemer** [comment]
Aria-owns is supported by firefox, Chrome,  as well (it actually changes the order of the accessibility tree). @jcsteh uses this in his grease monkey script for github quite extensively.
Here is a test which shows that the order of elements being owned is changed.
https://codepen.io/anon/pen/vrzvBg
This works in both firefox and Chrome.  Reordering like this currently does not seem to be supported by edge.

The following example is adapted from [1]. 
https://codepen.io/anon/pen/gKdZMX

This currently works in both Chrome and Firefox. Unfortunately, it seems that Edge does not support aria-owns, at least the accessibility tree is not reflecting those changes. While I can query the AriaProperties attribute on the UIA element in edge and see that it labels owns:child properly, it isn't updating the actual parent-child relationship at all. This is arguably a bug in Edge, not NVDA.

Secondly, when I closed this it was because our wiki states the following. 
> Warning: In all but exceptional circumstances we require this template to be completed. Your issue will likely be closed if this template has not been followed. [2] 

We have to follow a process to ensure we are triaging issues effectively as a small project with a mostly volunteer issue triage process. This issue was not a complete enough issue for us to do anything useful with in terms of meaningful impact without specific instructions on what in the spec was not implemented.

[2]: https://github.com/nvaccess/nvda/wiki/Github-issue-template-explanation-and-examples
[1]: https://tink.uk/using-the-aria-owns-attribute/

**Comment by jcsteh** [comment]
aria-owns is definitely supported by Firefox. This is something that needs to be supported by the browser (especially for Edge where we can't render virtual buffers). If Edge doesn't support this (and I can't comment on specifics of the implementation, since i haven't tested it), that is definitely a deficiency in Edge and cannot be fixed in NVDA.
